src/safeposix/syscalls/fs_calls.rs

```diff
@@ -267,76 +270,66 @@ impl Cage {
-                                if let Inode::Dir(ref mut parentdirinodeobj) = *parentinodeobj {
-                                    parentdirinodeobj.filename_to_inode_dict.insert(filename, inodenum);
-                                    parentdirinodeobj.linkcount += 1;
-                                    drop(parentinodeobj);
-                                    let newsockaddr = NET_METADATA.domain_socket_table.get(&trueoldpath).unwrap().clone();
-                                    NET_METADATA.domain_socket_table.insert(truenewpath, newsockaddr);
-                                } //insert a reference to the inode in the parent directory
-                                0 //link has succeeded
-                            }
+                    }
 
-                            (Some(_), ..) => {syscall_error(Errno::EEXIST, "link", "newpath already exists")}
+                    Inode::Dir(_) => {return syscall_error(Errno::EPERM, "link", "oldpath is a directory")}
+                }
+
+                drop(inodeobj);
+
+                let retval = match metawalkandparent(truenewpath.as_path()) {
+                    (None, None) => {syscall_error(Errno::ENOENT, "link", "newpath cannot be created")}
+
+                    (None, Some(pardirinode)) => {
+                        let mut parentinodeobj = FS_METADATA.inodetable.get_mut(&pardirinode).unwrap();
+                        //insert a reference to the inode in the parent directory
+                        if let Inode::Dir(ref mut parentdirinodeobj) = *parentinodeobj {
+                            parentdirinodeobj.filename_to_inode_dict.insert(filename, inodenum);
+                            parentdirinodeobj.linkcount += 1;
+                            drop(parentinodeobj);
+                            log_metadata(&FS_METADATA, pardirinode);
+                            log_metadata(&FS_METADATA, inodenum);
+                        } else {
+                            panic!("Parent directory was not a directory!");
                         }
+                        0 //link has succeeded
                     }
 
-                    Inode::Dir(_) => {syscall_error(Errno::EPERM, "link", "oldpath is a directory")}
+                    (Some(_), ..) => {syscall_error(Errno::EEXIST, "link", "newpath already exists")}
+                };
+
+                if retval != 0 {
+                    //reduce the linkcount to its previous value if linking failed
+                    let mut inodeobj = FS_METADATA.inodetable.get_mut(&inodenum).unwrap();
+
+                    match *inodeobj {
+                        Inode::File(ref mut normalfile_inode_obj) => {
+                            normalfile_inode_obj.linkcount -= 1;
+                        }
+
+                        Inode::CharDev(ref mut chardev_inode_obj) => {
+                            chardev_inode_obj.linkcount -= 1;
+                        }
+
```

Comment:
I think this extra newline is not needed.

Reply:
Fixed